flink-streaming-java/src/main/java/org/apache/flink/streaming/api/environment/StreamExecutionEnvironment.java

```diff
@@ -2092,6 +2094,27 @@ public StreamGraph getStreamGraph(String jobName) {
     @Internal
     public StreamGraph getStreamGraph(String jobName, boolean clearTransformations) {
         StreamGraph streamGraph = getStreamGraphGenerator().setJobName(jobName).generate();
+
+        // There might be a resource deadlock when applying fine-grained resource management in
+        // batch jobs with PIPELINE edges. Users need to trigger the
+        // fine-grained.shuffle-mode.all-blocking to convert all edges to BLOCKING before we fix
+        // that issue.
+        if (configuration.get(ExecutionOptions.RUNTIME_MODE) == RuntimeExecutionMode.BATCH
+                && !slotSharingGroupResources.isEmpty()) {
+            if (configuration.get(ClusterOptions.FINE_GRAINED_SHUFFLE_MODE_ALL_BLOCKING)) {
+                streamGraph.setGlobalDataExchangeMode(GlobalDataExchangeMode.ALL_EDGES_BLOCKING);
+            } else {
+                throw new IllegalStateException(
+                        "As described in FLINK-20865, there might be a resource deadlock when applying "
+                                + "fine-grained resource management in batch jobs with PIPELINE edges. "
+                                + "As a temporary workaround, please configure the "
+                                + ClusterOptions.FINE_GRAINED_SHUFFLE_MODE_ALL_BLOCKING.key()
+                                + " to true, with which Flink will convert those PIPELINE edges to BLOCKING. "
+                                + "Please not that data will be fully produced before sent to consumer tasks "
+                                + "for BLOCKING edges, that might affect the performance.");
+            }
```

Comment:
1. `IllegalStateException` -> `IllegalConfigurationException`
2. Not sure if we want to explain as detail as how blocking edges work. The error message should probably focus on "what is wrong" and "how to fix it".

> At the moment, fine-grained resource management requires batch workloads to be executed with types of all edges being BLOCKING. To do that, you need to configure '<config-key>' to 'true'. Notice that this may affect the performance. See FLINK-20865 for more details.